sor/src/main/java/com/bazaarvoice/emodb/sor/core/DefaultDataStore.java

```diff
@@ -368,12 +373,18 @@ private Resolved resolve(Record record, ReadConsistency consistency) {
     private Expanded expand(final Record record, boolean ignoreRecent, final ReadConsistency consistency) {
         long fullConsistencyTimeStamp = _dataWriterDao.getFullConsistencyTimestamp(record.getKey().getTable());
         long rawConsistencyTimeStamp = _dataWriterDao.getRawConsistencyTimestamp(record.getKey().getTable());
-        return expand(record, fullConsistencyTimeStamp, rawConsistencyTimeStamp, ignoreRecent, consistency);
+        Map<String, StashRunTimeInfo> stashTimeInfoMap = _compactionControlSource.getStashTimesForPlacement(record.getKey().getTable().getAvailability().getPlacement());
+        // we will consider the earliest timestamp found as our compactionControlTimestamp.
+        // we are also filtering out any expired timestamps. (CompactionControlMonitor should do this for us, but for now it's running every hour. So, just to fill that gap, we are filtering here.)
+        // If no timestamps are found, then taking minimum value because we want all the deltas after the compactionControlTimestamp to be deleted as per the compaction rules as usual.
+        long compactionControlTimestamp = stashTimeInfoMap.isEmpty() ?
+                Long.MIN_VALUE : stashTimeInfoMap.values().stream().filter(s -> s.getExpiredTimestamp() > System.currentTimeMillis()).map(StashRunTimeInfo::getTimestamp).min(Long::compareTo).get();
```

Comment:
There may be a problem here if `stashTimeInfoMap.isEmpty() == false` but every value in it is expired.  This will result in an empty map and calling `.get()` at the end would throw a `NoSuchElementException`.  Better to replace the final call with `.orElse(Long.MIN_VALUE)`.